### ISSUE TITLE
AttributeError("'bytes' object has no attribute 'encode'",)

### ISSUE THREAD

**Issue by astronouth7303** [celery/celery#3457]
Running Celery 4.0.0rc4 on Python 3.4.3 on AWS EC2:

```
Traceback (most recent call last):
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/kombu/utils/objects.py", line 40, in __get__
    return obj.__dict__[self.__name__]
KeyError: 'scheduler'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.4/shelve.py", line 111, in __getitem__
    value = self.cache[key]
KeyError: b'entries'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/apps/beat.py", line 107, in start_scheduler
    service.start()
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/beat.py", line 522, in start
    humanize_seconds(self.scheduler.max_interval))
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/kombu/utils/objects.py", line 42, in __get__
    value = obj.__dict__[self.__name__] = self.__get(obj)
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/beat.py", line 566, in scheduler
    return self.get_scheduler()
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/beat.py", line 561, in get_scheduler
    lazy=lazy,
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/beat.py", line 406, in __init__
    Scheduler.__init__(self, *args, **kwargs)
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/beat.py", line 205, in __init__
    self.setup_schedule()
  File "/opt/python/run/venv/local/lib/python3.4/site-packages/celery/beat.py", line 436, in setup_schedule
    self._store[b'entries']
  File "/usr/lib64/python3.4/shelve.py", line 113, in __getitem__
    f = BytesIO(self.dict[key.encode(self.keyencoding)])
AttributeError: 'bytes' object has no attribute 'encode'
```

**Comment by astronouth7303** [comment]
My applicable settings:

```
CELERY_RESULT_BACKEND = 'django-cache'
CELERY_BROKER_URL = 'sqs://'
CELERY_BROKER_TRANSPORT_OPTIONS = {
    'queue_name_prefix': os.environ.get('AWS_STACK_NAME', 'test') + '-',
}

CELERY_BEAT_SCHEDULE = {
    
}
```

**Comment by ask** [comment]
It seems you used the previous rc, so you need to remove your celerybeat-schedule file now! Sorry for that.

**Comment by astronouth7303** [comment]
Nope. I've been using 4.0.0rc4 for the entire lifetime of this project. In addition, I'm on AWS ElasticBeanstalk; I've been knocking over and spinning up new servers regularly.

**Comment by ask** [comment]
Ok, so then I guess shelve on Python 3.4 doesn't support bytes...

**Comment by ask** [comment]
but possibly we could set that keyencoding attribute

**Comment by jchappell82** [comment]
As additional information, I get this exact same error on Python 3.5.2 as well. (in this case running on macOS Sierra)

I am similarly using SQS as the broker, and this was a fresh start with 4.0.0rc4, not an upgrade from a previous version.

**Comment by mjmonroe** [comment]
It appears to be fixed in the https://github.com/celery/celery/commit/d87f9db7371838369e7ff6cbb400d29f9220a789 commit.

**Comment by astronouth7303** [comment]
As weird of construction as that is (`str(b'...')`? Really?), it seems to work.

Looking forward to the next RC.

**Comment by astronouth7303** [comment]
Yeah, confirmed. This is cleared up on master.

**Comment by ggstuart** [comment]
Hi, I'm on 3.1.24 from pip and am getting this using python 3.4 on both centOS 7 and ubuntu 14.04. 

Is the recommended approach to upgrade to the latest release candidate?

**Comment by mjmonroe** [comment]
@ggstuart Celery 4.0.0rc4 has the bug, you need to upgrade to at least https://github.com/celery/celery/commit/d87f9db7371838369e7ff6cbb400d29f9220a789 commit. Pip can be used to install from a[ github commit](https://pip.pypa.io/en/latest/reference/pip_install/#git).

**Comment by ask** [comment]
Celery 4.0.0rc5 will be up on PyPI in a few minutes

**Comment by ask** [comment]
and Celery 4.0.0rc5 is on PyPI :tada: 

Note that this depends on Kombu 4.0.0rc5, which was also uploaded just now, but pip does
not let us specify dependencies for dev versions, so if it crashes then try to upgrade kombu manually.

Use:

```console

$ pip install 'celery>dev'

```

**Comment by alzeih** [comment]
@astronouth7303 @ask Ha! I overlooked just str('') 👎 

This is why just a literal doesn't work though:
``` 
python3 -c "from __future__ import unicode_literals; print(type('foo'))"
<class 'str'>
python3 -c "from __future__ import unicode_literals; print(type(b'foo'))"
<class 'bytes'>
python3 -c "from __future__ import unicode_literals; print(type(u'foo'))"
<class 'str'>
python2 -c "from __future__ import unicode_literals; print(type('foo'))"
<type 'unicode'>
python2 -c "from __future__ import unicode_literals; print(type(b'foo'))"
<type 'str'>
python2 -c "from __future__ import unicode_literals; print(type(u'foo'))"
<type 'unicode'>
```

**Comment by FirefighterBlu3** [comment]
this doesn't look to be fixed in the release version 4.0.0 (i'm using py3).

```python
   File "/usr/lib/python3.5/site-packages/celery/worker/state.py", line 232, in _merge_revoked
     self._merge_revoked_v3(d[b'zrevoked'])
   File "/usr/lib/python3.5/shelve.py", line 113, in __getitem__
     f = BytesIO(self.dict[key.encode(self.keyencoding)])
 AttributeError: 'bytes' object has no attribute 'encode'
```

**Comment by alzeih** [comment]
Looks like some bytestrings added in https://github.com/celery/celery/commit/356648ab71193ffeecba940ba23deca27a53976d and https://github.com/celery/celery/commit/0d90949cd4e033f5522ca1e15e9bf13ef02709be

I have a patch, will send PR shortly
